src/notify/notify.c

```diff
@@ -207,12 +216,22 @@ static int run(int argc, char* argv[]) {
             setreuid(arg_uid, (uid_t) -1) < 0)
                 return log_error_errno(errno, "Failed to change UID: %m");
 
-        r = sd_pid_notify(arg_pid ? arg_pid : getppid(), false, n);
+        r = sd_pid_notify(arg_pid ? arg_pid : (arg_no_block ? getppid() : 0), false, n);
```

Comment:
hmm, why this? I think it is still better to return the ppid here, since we usually send this on behalf of the caller, not because we actually want to notify anyone ourselves

Reply:
We wouldn't need to do it without --no-block, and it never really worked all the time, but only when the caller had enough privs, so you cannot really rely on it unless you explicitly set it to yours using --pid= (which again will not work without privs); Though I can drop it ofcourse...

Reply:
There's also the problem where systemd-notify when invoked by PID 1 as root as a control process ends up sending PID=1, which is duly rejected by the manager.

A simple command:
`systemd-run --unit=foobar.service -p Type=notify -p NotifyAccess=all /bin/systemd-notify --ready`

fails without this change.

Reply:
So the summary is:
* Currently systemd-notify is also broken for cases where it has the privileges to attest the notification message using the parent PID and is invoked by the service manager. It will always fail, regardless of whether it raced or not.
* With this change, --no-block still tries the PPID trick, but that's the next best thing we can do after the user asked us to disable sd_notify_barrier() polling, so it might make sense keeping it, and I already changed the explanation for the PPID stuff to tell it is only done when --no-block is used in the manual.
* The PPID trick is quite sensible in case of --no-block, because if the calling control process has the right privileges, and wants to invoke systemd-notify on its behalf, it can do so without waiting (since its PID is mapped to the unit name in PID 1). However it is not so useful when the process is unprivileged, so in that case --no-block should be avoided.

Reply:
the way i see it there are two reasons for the getppid() hack: a) to avoid the race if possible, b) because this is a tool that is generally used by shell scripts, and in that case it's generally not systemd-notify itself that genuinely out of itself wants to send the message to the service manager, but it's doing so only on behalf, on request of the script it is invoked from. Hence, philosophically I think we should attempt to report who actually wants to send the msg here, i.e. the parent — if we can.

I mean, you do have a point, it's a bit ugly that we do this at all, but there's are reasons for it, and I think simpler rules (i.e. not also mixing blocking vs. non-blocking mode into the problem) are good here. Or to turn this around: i don't see why doing the ppid trick would hurt us here. i.e. yes, i think we should an extra tweak and disable the ppid trick if ppid happens to be 1, but i don't see how it could hurt in other cases...

Reply:
So I prepped #15642 now in response of this, since you correctly point out that we should never send messages from a faked source of PID 1. Could you have a look?